packages/openapi3/test/shared-routes.test.ts

```diff
@@ -0,0 +1,140 @@
+      @service({title: "My Service"})
+      namespace Foo {
+        model Resource {
+          id: string;
+        }
+
+        @route("/sharedroutes/resources?filter=resourceGroup")
+        op listByResourceGroup(...Resource): Resource[];
+
+        @route("/sharedroutes/resources?filter=subscription")
+        op listBySubscription(...Resource): Resource[];
+      }
+      `
+    );
+    expectDiagnostics(diagnostics, [
+      {
+        code: "@typespec/openapi3/path-query",
+        message: "OpenAPI does not allow paths containing a query string.",
+      },
+      {
+        code: "@typespec/openapi3/path-query",
+        message: "OpenAPI does not allow paths containing a query string.",
+      },
+    ]);
+  });
+
+  it("model shared routes that differ by query params as a union of optional params", async () => {
+    const results = await openApiFor(
+      `
+      @service({title: "My Service"})
+      namespace Foo {
+        model Resource {
+          id: string;
+        }
+
+        @route("/sharedroutes/resources", { shared: true })
+        op listByResourceGroup(...Resource, @query resourceGroup: string, @query foo: string): Resource[];
+
+        @route("/sharedroutes/resources", { shared: true })
+        op listBySubscription(...Resource, @query subscription: string, @query foo: string): Resource[];
+      }
+      `
+    );
+    const params = results.paths["/sharedroutes/resources"].post.parameters as {
+      name: string;
+      required: boolean;
+    }[];
+    params.sort((a, b) => a.name.localeCompare(b.name));
+    deepStrictEqual(params[0].name, "foo");
+    deepStrictEqual(params[0].required, true);
+    deepStrictEqual(params[1].name, "resourceGroup");
+    deepStrictEqual(params[1].required, false);
+    deepStrictEqual(params[2].name, "subscription");
+    deepStrictEqual(params[2].required, false);
+  });
+
+  it("model shared routes that differ by query param values as an enum", async () => {
+    const results = await openApiFor(
+      `
+      @service({title: "My Service"})
+      namespace Foo {
+        model Resource {
+          id: string;
+        }
+
+        @route("/sharedroutes/resources", { shared: true })
+        op listByResourceGroup(...Resource, @query filter: "resourceGroup"): Resource[];
+
+        @route("/sharedroutes/resources", { shared: true })
+        op listBySubscription(...Resource, @query filter: "subscription"): Resource[];
+      }
+      `
+    );
+    const params = results.paths["/sharedroutes/resources"].post.parameters as {
+      name: string;
+      required: boolean;
+      schema: any;
+    }[];
+    params.sort((a, b) => a.name.localeCompare(b.name));
+    deepStrictEqual(params[0].name, "filter");
+    deepStrictEqual(params[0].required, true);
+    deepStrictEqual(params[0].schema, {
+      type: "string",
+      enum: ["resourceGroup", "subscription"],
+    });
+  });
+
+  it("model shared routes with shared parameters in different locations", async () => {
```

Comment:
A couple of important scenarios to cover:
- ops on a shared route that have different return types
  ```typespec
  @route("/foo", {shared: true})
  op foint(@query options: string): int32;
  @route("/foo", {shared: true})
  op fooString(@query options: string): string;
  ```
 Should result in the response body using a oneOf
- ops on a shared route that have different request body types:
  ```typespec
  @route("/foo", {shared: true})
  @put op fooInt(@body body: int32): void;
  @route("/foo", {shared: true})
  @put op fooString(@body body: string): void;
  ```
(should result in the request body using a oneOf)
- combination of 1 and 2
  ```typespec
  
  @route("/foo", {shared: true})
  @action("upload") op fooInt(body: int32): int32;
  @route("/foo", {shared: true})
  @action("upload") op fooString(body: string): string;
  ```
  
result should have a one-of for the body parameter and the response

Reply:
Thanks @markcowl I added test cases for each of these scenarios.